include/LightGBM/bin.h

```diff
@@ -98,7 +98,7 @@ class BinMapper {
   * \brief Save binary data to file
   * \param file File want to write
   */
-  void SaveBinaryToFile(FILE* file) const;
+  void SaveBinaryToFile(std::shared_ptr<VirtualFileWriter> file) const;
```

Comment:
why use `share_ptr` ? we often use `const` raw pointers. what are the advantages of share_ptr ?

Reply:
It's only a personal preference, these objects are very low volume so the overhead should be minimal, and the shared_ptr made thinking about the lifetime much easier. If you prefer though i can re-work this so the base instance gets allocated on the stack and we pass const pointers. Stack allocation should keep the relatively simple lifetime.

Reply:
okay, got it. The pointer is actual not 'const' inside these functions.